### PR TITLE
- added the ability to only download  binary dependencies (#2229)

### PR DESCRIPTION
**Pull request by KoCMoHaBTa**

### CODE REVIEW

**Comment by KoCMoHaBTa** [comment]
@mdiep What would be the desired behaviour of Carthage when it encounters a static framework?
Is it only identifying its platform and putting it inside the `Carthage/Build/<platform>` folder?

**Comment by KoCMoHaBTa** [comment]
@mdiep Added support for static frameworks during `checkout`.

File: Source/carthage/Checkout.swift [hunk 1]
```diff
@@ -46,6 +52,8 @@ public struct CheckoutCommand: CommandProtocol {
 				<*> mode <| Option(key: "use-ssh", defaultValue: false, usage: "use SSH for downloading GitHub repositories")
 				<*> mode <| Option(key: "use-submodules", defaultValue: false, usage: "add dependencies as Git submodules")
 				<*> mode <| Option(key: "use-binaries", defaultValue: true, usage: useBinariesUsage)
+				<*> mode <| Option(key: "copy-binaries", defaultValue: false, usage: "copy the contents of the binary depndencies")
+				<*> mode <| Option(key: "copy-frameworks", defaultValue: true, usage: "skip frameworks validations and copying in Build folder, can be used with --copy-binaries in order to only download raw binary dependencies")
```
Comment (mdiep) [inline Source/carthage/Checkout.swift]: These options should be removed.

File: Source/CarthageKit/Project.swift [hunk 2]
```diff
@@ -1087,8 +1112,41 @@ private func platformForFramework(_ frameworkURL: URL) -> SignalProducer<Platfor
 				let error = Result<(), NSError>.error(message)
 				return .readFailed(frameworkURL, error)
 			}
+			
+			func sdknameFromExecutable() -> Any? {
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: The _n_ in _name_ should be capitalized here.

File: Source/CarthageKit/Constants.swift [hunk 3]
```diff
@@ -9,6 +9,10 @@ public struct Constants {
 	/// The name of the folder into which Carthage puts binaries it builds (relative
 	/// to the working directory).
 	public static let binariesFolderPath = "Carthage/Build"
+	
+	/// The name of the folder into which Carthage puts binaries it downloads (relative
+	/// to the working directory).
+	public static let downloadsFolderPath = "Carthage/Downloads"
```
Comment (mdiep) [inline Source/CarthageKit/Constants.swift]: This should be removed.

File: Source/CarthageKit/Project.swift [hunk 4]
```diff
@@ -104,6 +104,12 @@ public final class Project { // swiftlint:disable:this type_body_length
 	/// Whether to download binaries for dependencies, or just check out their
 	/// repositories.
 	public var useBinaries = false
+	
+	/// Whenever to copy the downloaded binary dependencies
+	public var copyBinaries: Bool = false
+	
+	/// Whenever to validate and copy the downloaded frameworks to the Build directory
+	public var copyFrameworks: Bool = true
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: These options should be removed.

**Comment by KoCMoHaBTa** [comment]
@mdiep I don't understand why you won't accept the feature to expose the contents of the downloaded binary dependency as an option.

What happened with the Carthage statement:

> Carthage builds framework binaries using xcodebuild, but leaves the responsibility of integrating them up to the user.

In this case - the framework binaries are prebuilt, but we don't have the ability to integrate them by ourselves.
**It is not Carthage responsibility to inspect, validate and group prebuilt framework binaries.**
It responsibility should be only to resolve them.

**Comment by blender** [comment]
Please add appropriate tests

**Comment by mdiep** [comment]
> I don't understand why you won't accept the feature to expose the contents of the downloaded binary dependency as an option.

It's not necessary. If a `binary` origin can copy the pre-built static framework to the appropriate `Carthage/Build/` directory, then why do you also need to expose the contents of the zip in `Carthage/Downloads/`?

**Comment by KoCMoHaBTa** [comment]
Mainly, because of:
- the resources that comes with the static library or framework - how would you handle it?
- specific instructions that you might need to follow from the provider of the library
- this will give you the ability to integrate with any kind of dependency for free. Why would you have this limitation?

The `Carthage/Build` folder is intended to put build that Carthage has just builded itself.
Binary dependency means - as is. Carthage should not even try to assume what it is - because it really can be anything.

From the beginning i'm talking about this. 
Since i've wasted too much time, trying to convince you that Carthage can do much more by support anything as dependency with this small addition, i'm considering to move forward.

**So if this is going to be accepted as an idea** - i'm keen to only do changes related to configurations and conventions. Otherwise, just let me know and i'll no longer bother you, as obviously the acceptance of good ideas appears to be very conservative for this repository.

I believe i've made myself, my idea and my reasons clear few times so far. I prefer to do something useful rather than continue spending time arguing.

**Comment by mdiep** [comment]
> the resources that comes with the static library or framework - how would you handle it?

This question doesn't make sense for a static framework—these resources should be bundled _in_ the static framework. Copying that framework gets you the resources.

The solution I've suggested meets the need that you've presented (a vendor-provided static framework) and is consistent with the overall approach for Carthage. If there's another need that it's not meeting, then we should discuss that explicitly.

**Comment by KoCMoHaBTa** [comment]
Static frameworks cannot contain resources, because they are static libraries wrapped to look like frameworks. 
Unfortunately a lot of big companies continue to use this approach - https://github.com/jverkoey/iOS-Framework
And for resources, as these are actually static libraries - they cannot be embedded inside, but are distributed separately, most of the time as bundles, but not always - https://github.com/jverkoey/iOS-Framework#resources

**Comment by mdiep** [comment]
> Static frameworks cannot contain resources, because they are static libraries wrapped to look like frameworks.

Yes they can. ☺️ A static framework is just a special folder with the static library inside. You can put whatever you like inside of it.

But it sounds like that's not how people are actually doing it.

**Comment by KoCMoHaBTa** [comment]
Closing due to inability and unwillingness of the owner to understand the proposed feature and benefits of supporting any kind of binary dependency.

**Pull request closed without merging**